### PR TITLE
Add more ES6 features to `.eslintrc` and remove `babel-eslint`

### PR DESCRIPTION
**Pull request by nkbt**
Separate PR for `babel-eslint` replace with ESLint's own `ecmaFeatures`.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -8,10 +6,28 @@
     },
 
     "ecmaFeatures": {
-        "jsx": true,
-        "superInFunctions": false,
+        "arrowFunctions": true,
+        "binaryLiterals": true,
+        "blockBindings": true,
         "classes": false,
-        "modules": [2]
```
Comment (nkbt) [inline .eslintrc]: The syntax for ecmaFeatures is `true/false` not as for rules.

**Comment by ericelliott** [comment]
Still pulling out `babel-eslint`? We want to disallow ES7 features like object rest/spread?

And does babel-eslint require the `ecmaFeatures` at all?

**Comment by nkbt** [comment]
As I mentioned, we use `spread` and `rest` and some other pretty modern features in our main pretty large project without babel-eslint with no problems. Everything is es-linted correctly. moreover, using `babel-eslint` caused problems due to slightly altered formatting which was not compatible with our strict code-style config.

So far I have not experienced any need in it.

`ecmaFeatures` are needed only for ESLint itself, not `babel-eslint`, which just transforms the code before linting.

**Comment by ericelliott** [comment]
> As I mentioned, we use `spread` and `rest` and some other pretty modern features in our main pretty large project without babel-eslint with no problems.

I'm going to go out on a limb and guess that you're using **array spread** and **array rest** forms (params, arguments and arrays), because ESLint doesn't recognize **object rest** and **object spread** as valid syntax.

> Everything is es-linted correctly. moreover, using `babel-eslint` caused problems due to slightly altered formatting which was not compatible with our strict code-style config.

What problems, and whose code-style config?

> So far I have not experienced any need in it.

That doesn't mean they don't exist.

**Comment by nkbt** [comment]
Right. Your bleeding-edge-fu is stronger then mine!

By our code-style config I meant the one we use in the company I work with, sorry for confusion.

**Comment by ericelliott** [comment]
> By our code-style config I meant the one we use in the company I work with, sorry for confusion.

Is there anything we can do to make our ESLint config work for your code using `babel-eslint`?

Remember ESLint is extensible and open-source, so if there are rules you need (that make sense for this module's intended use-cases), maybe we could just reproduce them?

**Comment by nkbt** [comment]
I put babel-eslint back, but there is one more thing to consider.

https://github.com/sebmarkbage/ecmascript-rest-spread
I reckon this is not supported by babel without additional flags.



So either way if you want to use them (and other draft/proposal features), npm build script has to be updated anyway.

**Comment by nkbt** [comment]
By the way if you meant object spread like
```js
<Component {...this.props} />
```

It is used in my code and does not require `babel-eslint`.

**Comment by ericelliott** [comment]
That syntax is in the context of JSX, which has explicit support in ESLint. Try it in a POJO object literal. If that works, we can drop `babel-eslint`.

If it doesn't, we need to keep it and add the config to enable the features in babel.

**Comment by nkbt** [comment]
Apparently it is enabled for JSX only, not for POJO



So let's keep `babel-eslint` for a moment (reverted by second commit).

Enabling experimental features is something to discuss if it is reasonable to have them in `prod-module`.

**Comment by ericelliott** [comment]
Yeah, normally I'd be more conservative about experimental features, but object rest/spread are being used in a lot of React documentation. It may be too late to put that genie back in the bottle.

**Comment by ericelliott** [comment]
Babel says it [supports object rest/spread on the homepage](https://babeljs.io/). Let's make sure it's working in our repo before we merge this. =)

**Comment by ericelliott** [comment]
Status? Is it safe to merge this? I haven't tested... o.O

**Comment by nkbt** [comment]
I actually put back babel-eslint, since object rest/spread does not work out of box, so it is absolutely safe to merge (now it just has a complete list of ecmaFeatures in eslint config).

**Comment by nkbt** [comment]
:+1:

**Pull request merged**